bsip-0043.md

```diff
@@ -0,0 +1,96 @@
+The workflow may be as following:
+
+An asset issuer defines the market_fee_reward_percent - what percentage of the market fee he wants to share with the registrar in asset options. Registrar defines reward_percent for the referrer for each user(using the already existing BitShares mechanism).
+Market fee reward is accumulated on the user account. The user decides when they want to claim the market fee reward and move it to their active balance. There is another operation created asset_claim_reward_operation. Each user pays network fee to call this operation.
+
+# Motivation
+
+To make promoting BitShares and bringing new users much more attractive to registrars and referrers by sharing UIAs and MPAs market fees with them.
+
+# Rationale
+When 'fill_order_operation' executing at the moment of market fee calculation there will be calculate the reward for the registrar used the parameter market_fee_reward_percent. Then this market_fee_reward_percent of the market fee is split between the registrar and referrer according to the referrer_rewards_percentage, which is set up during the new account registration by registrar (please see the parameters for register_account).
+
+# Specifications
+
+## market_fee_reward_percent asset option
+Percent of market fee that will be paid to buyer's registrar. Set by UIA issuer.
+
+## market_sharing_whitelist asset option
+An optional list of accounts (configurable by the UIA Issuer) could provide increased control over who is eligible to get market rewards.
+
+## New database and wallet api methods
+**get_account_reward(account, asset_symbol)** - Returns information about the reward amount for specific asset.
+**list_account_rewards(account)** - Returns information about the reward amount in various assets
+**claim_reward(account, asset_symbol, amount)** - Claim account's reward for the given asset.
+
+## account_reward_object
+A new BitShares object type in implementation space impl_account_reward_object_type = 2.18.X that tracks the rewards of a single account/asset pair
+```
+account_reward_object {
+owner_account,
+asset_type,
+reward_amount
+}
+```
+## account_reward_index
+A new index that stores objects of account_reward_object-type in graphene::database and allow random fast access to objects by given criteria
+```
+account_reward_index multi-index of account_reward_objects
+indexed_by
+[id]
+[owner_account, asset_type]
+[asset_type, reward_amount desc, owner_account]
+```
+## asset_claim_reward_operation
+A new operation used to transfer reward to the account's balance.
```

Comment:
Please explain why this is required.
It would be much simpler to either pay these rewards automatically, e. g. during maintenance like the current rewards system does, or to use the existing vesting balance mechanism.

Reply:
concerning vesting balance mechanism, please see our response to xeroc above and to abit below.
And paying out directly during maintenance, if we understand correctly, there is no such mechanism right now. In maintenenance, only process_fees is called, that distributes accumulated fees to the network, registrar and rererrer using deposit_cashback (the vesting balance mechanism we discussed above).

Reply:
> see our response to xeroc above and to abit below.

Sorry I didn't see a new response so far.

> ... if we understand correctly, there is no such mechanism right now. In maintenenance, only process_fees is called ...

Since we're discussing how to implement a new feature, I think adding more tasks to be done in maintenance interval **IS** an option.

Reply:
>Sorry I didn't see a new response so far.

sorry, @abitmore we planned to respond to your commment below, but i think we covered most of it to the response to Peter's comment above :)
Denis

Reply:
> why fees distribution happens through vesting balance.

1. cashback is designed to have a vesting period;
2. for better performance for non-maintenance blocks (only process fees on maintenance blocks). Updating one object v.s. updating 3 objects and related looking-ups for the objects. Arguable though.

Reply:
Guys, we had another discussion today about direct transfer.  So we think having direct transfer does not generate separate transactions and hence does not generate network fee. 
So we think that the best way would be to recreate  vesting-like mechanism - so that we utilize  a similar mechanism to the  current mechanism of claiming accumulated fees from assets  (but for referrers and registrars. ) Then they have to use claim  - to claim their accumulated fees and hence, generate network fees for bitshares.

Alternatively, as we mentioned, using current vesting mechanism would be a great idea, but we then need to discuss the estimation and priority for #1276. Should we start this discussion? 
What do you think?
Denis

Reply:
Sorry, I didn't get it. Which is better in your opinion, "generate network fees", or not? I foresee that reward claiming fee won't be any big part of overall transaction fee. It's even possible that the committee will set balance claiming fee to zero since it removes data from RAM.

From a technical point of view, having to claim means more signed transaction on the chain, which will lead to chain bloating to an extent (more permanent storage); having to maintain unclaimed balances means larger chain state (more RAM usage). On the other hand, seems it requires less computation (less CPU usage).

In regards to user experience (UX) and business development (BD), each approach has its own pros and cons.

It's all about compromise.

Reply:
I want to add that it also makes a differences for taxes.

In germany, vesting balance is an "Option", whereas a direct payout would have direct tax implications.